[Airflow 16364] Add conn_timeout and cmd_timeout params to SSHOperator; add conn_timeout param to SSHHook

fatmumuhomer opened the pull request:
closes #16364 (Timeout is ambiguous in SSHHook and SSHOperator)

Add conn_timeout and cmd_timeout params to SSHOperator. Add conn_timeout param to SSHHook. The legacy timeout param remains but is marked as deprecated. The new params take precedence over the legacy timeout one if both/all are provided.

---
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

fatmumuhomer committed:
Issue 16363 - add conn_timeout and cmd_timeout params to SSHOperator; add conn_timoue param to SSHHook

fatmumuhomer committed:
Issue 16363 - add conn_timeout and cmd_timeout params to SSHOperator; add conn_timoue param to SSHHook

fatmumuhomer committed:
Issue 16363 - add conn_timeout and cmd_timeout params to SSHOperator; add conn_timoue param to SSHHook

fatmumuhomer committed:
Issue 16363 - add conn_timeout and cmd_timeout params to SSHOperator; add conn_timoue param to SSHHook

fatmumuhomer committed:
Update SSHOperator tests for new conn_timeout and cmd_timeout params

fatmumuhomer committed:
Update SSHOperator tests for new conn_timeout and cmd_timeout params

fatmumuhomer committed:
Update SSHOperator tests for new conn_timeout and cmd_timeout params

fatmumuhomer committed:
Update SSHOperator tests for new conn_timeout and cmd_timeout params

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Merge branch 'airflow-16364' of https://github.com/fatmumuhomer/airflow into airflow-16364

fatmumuhomer committed:
Removed commented TODO for SSHHook changes

fatmumuhomer committed:
Removed commented TODO for SSHHook changes

fatmumuhomer committed:
Removed commented TODO for SSHHook changes

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Fix static checks for operators/test_ssh.py and hooks/test_ssh.py

fatmumuhomer committed:
Fix static checks for operators/test_ssh.py and hooks/test_ssh.py

fatmumuhomer committed:
Fix static checks for operators/test_ssh.py and hooks/test_ssh.py

fatmumuhomer committed:
Merge branch 'airflow-16364' of https://github.com/fatmumuhomer/airflow into airflow-16364

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

uranusjr commented on airflow/providers/ssh/hooks/ssh.py:
```diff
@@ -56,8 +56,10 @@ class SSHHook(BaseHook):
+    :param timeout: (Deprecated). timeout for the attempt to connect to the remote_host.
     :type timeout: int
+    :param conn_timeout: timeout for the attempt to connect to the remote_host.
+    :type conn_timeout: int
```
We should put this above the deprecated `timeout`

uranusjr commented on airflow/providers/ssh/hooks/ssh.py:
```diff
@@ -101,7 +103,8 @@ def __init__(
         key_file: Optional[str] = None,
         port: Optional[int] = None,
-        timeout: int = 10,
+        timeout: Optional[int] = 10,
```
Is this change needed/correct? I feel we want to either change the default to `None` (so we can better detect whether `timeout` is passed in by the user), or the type should continue to be `int` (since the user shouldn’t do `timeout=None` in any situation).

fatmumuhomer replied:
Yeah, looking at it again you're right. It should default to None otherwise the user will need to purposefully pass timeout = None to turn off the deprecation warning.

In order to maintain backwards compatibility, since the current versions of the hook/operator allow the user to omit timeout, I'll need to handle a situation where no timeout is passed and set the default later in the code.

uranusjr commented on airflow/providers/ssh/hooks/ssh.py:
```diff
@@ -185,6 +200,18 @@ def __init__(
                     self.host_key = key_constructor(data=decoded_host_key)
                     self.no_host_key_check = False
 
+        if self.timeout:
```
Continuing above, this check seems to make more sense if the default `timeout` is `None`.

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Merge branch 'airflow-16364' of https://github.com/fatmumuhomer/airflow into airflow-16364

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Set deprecated timeout param default value to None; add more tests

fatmumuhomer committed:
Set deprecated timeout param default value to None; add more tests

fatmumuhomer committed:
Merge branch 'airflow-16364' of https://github.com/fatmumuhomer/airflow into airflow-16364

uranusjr commented on airflow/providers/ssh/hooks/ssh.py:
```diff
@@ -148,7 +155,17 @@ def __init__(
+                        DeprecationWarning,
+                        stacklevel=2,
+                    )
+                    self.timeout = int(extra_options.get('timeout', 10))
```
```suggestion
                    self.timeout = int(extra_options['timeout'])
```

We already know `"timeout"` is in `extra_options` (the `if` check above), this `get()` is not needed.

uranusjr commented on airflow/providers/ssh/hooks/ssh.py:
```diff
@@ -148,7 +155,17 @@ def __init__(
+                    self.timeout = int(extra_options.get('timeout', 10))
+
+                if "conn_timeout" in extra_options:
+                    self.conn_timeout = int(extra_options.get('conn_timeout'))
```
```suggestion
                    self.conn_timeout = int(extra_options['conn_timeout'])
```

uranusjr reviewed (changes requested):
Almost there!

fatmumuhomer committed:
Remove unnecessary extra_options.get() calls for timeout and conn_timeout

fatmumuhomer committed:
Remove unnecessary extra_options.get() calls for timeout and conn_timeout

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

jedcunningham commented on airflow/providers/ssh/hooks/ssh.py:
```diff
@@ -56,7 +58,9 @@ class SSHHook(BaseHook):
-    :param timeout: timeout for the attempt to connect to the remote_host.
+    :param conn_timeout: timeout for the attempt to connect to the remote_host.
+    :type conn_timeout: int
+    :param timeout: (Deprecated). timeout for the attempt to connect to the remote_host.
```
Should we mention what param to use instead?

potiuk replied:
Yeah I think it is a bit ambiguous now. with both params in the docs :)

jedcunningham commented on airflow/providers/ssh/operators/ssh.py:
```diff
@@ -43,7 +46,11 @@ class SSHOperator(BaseOperator):
+    :type conn_timeout: int
+    :param cmd_timeout: timeout (in seconds) for executing the command. The default is 10 seconds.
+    :type cmd_timeout: int
+    :param timeout: (deprecated) timeout (in seconds) for executing the command. The default is 10 seconds.
```
Here as well?

jedcunningham commented on tests/providers/ssh/hooks/test_ssh.py:
```diff
@@ -475,6 +504,120 @@ def test_ssh_connection_with_no_host_key_where_no_host_key_check_is_false(self,
+            ssh_mock.return_value.connect.assert_called_once_with(
+                hostname='remote_host',
+                username='username',
+                timeout=20,
```
I realize the existing behavior was for extras to win, but that seems backwards to me. I'd expect the param to win.

Maybe we leave `timeout` as-is, but invert it for `conn_timeout` since it is new?

(`hostname` is a good example)

fatmumuhomer replied:
I'm in favor of that behavior change. I agree - it would make sense and be more transparent if the explicit parameters won. I'm happy to make the change. Any objections, @potiuk?

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Add reference to new timeout parameters in docstrings for deprecated timeout

fatmumuhomer committed:
Add reference to new timeout parameters in docstrings for deprecated timeout

fatmumuhomer committed:
Move timeout docstring text to 2nd line to keep line within 110 characters

fatmumuhomer committed:
Move timeout docstring text to 2nd line to keep line within 110 characters

fatmumuhomer committed:
Favor conn_timeout parameter over extra options

fatmumuhomer committed:
Favor conn_timeout parameter over extra options

jedcunningham reviewed (changes requested):
Overall LGTM. One more thing I just thought of - we need to add `conn_timeout` and mark `timeout` as deprecated in the docs for the connection:
https://github.com/apache/airflow/blob/main/docs/apache-airflow-providers-ssh/connections/ssh.rst

Might not be a bad idea to also specify whether it takes precedence over the hook param too.

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Add conn_timeout to SSH Connection docs and mark timeout as deprecated

fatmumuhomer committed:
Add conn_timeout to SSH Connection docs and mark timeout as deprecated

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

jedcunningham commented on tests/providers/ssh/hooks/test_ssh.py:
```diff
@@ -475,6 +504,122 @@ def test_ssh_connection_with_no_host_key_where_no_host_key_check_is_false(self,
+                port='port',
+                sock=None,
+                look_for_keys=True,
+            )
```
Aren't we missing a test where neither `timeout` and `conn_timeout` are passed and we use the value from the connection extras? Also missing a test where the default is used if the connection doesn't have one set?

fatmumuhomer replied:
I was thinking about this a bit more - within SSHHook there are 16 ways to combine the two parameters and the two extra options. To be safe, the tests should really check them all to ensure we're selecting the value we expect each time. I'm working on a parametrized version that covers all 16 combinations.

jedcunningham replied:
Cool, sounds good 👍

fatmumuhomer replied:
The parametrized tests are done and I removed the non-parametrized ones that were now duplicates.

jedcunningham commented on airflow/providers/ssh/operators/ssh.py:
```diff
@@ -77,9 +87,24 @@ def __init__(
+        self.conn_timeout = conn_timeout
+        self.cmd_timeout = cmd_timeout
+        if self.conn_timeout is None:
+            self.conn_timeout = self.timeout if self.timeout else TIMEOUT_DEFAULT
```
I think by setting a default here, we effectively break allowing the connections `conn_timeout` to be used over the default. Maybe this would be better? Btw, I think we need better test coverage on default / extra handling like the hook too.

```suggestion
        if self.conn_timeout is None and self.timeout:
            self.conn_timeout = self.timeout
```

This'll let `None` be passed into the hook, which will let the connection extras be used and we avoid having the "default" set in 2 places.

fatmumuhomer replied:
Good catch! A test around this case would have caught this, too, so my fault for not creating one. I'll tackle both of those tonight.

jedcunningham commented on airflow/providers/ssh/operators/ssh.py:
```diff
@@ -43,7 +46,12 @@ class SSHOperator(BaseOperator):
     :param command: command to execute on remote host. (templated)
     :type command: str
-    :param timeout: timeout (in seconds) for executing the command. The default is 10 seconds.
+    :param conn_timeout: timeout (in seconds) for maintaining the connection. The default is 10 seconds.
```
Maybe the same language here as is used on the `remote_host`? "If provided, it will replace..."

jedcunningham commented on docs/apache-airflow-providers-ssh/connections/ssh.rst:
```diff
@@ -47,7 +47,8 @@ Extra (optional)
     * ``private_key`` - Content of the private key used to connect to the remote_host.
     * ``private_key_passphrase`` - Content of the private key passphrase used to decrypt the private key.
-    * ``timeout`` - An optional timeout (in seconds) for the TCP connect. Default is ``10``.
+    * ``conn_timeout`` - An optional timeout (in seconds) for the TCP connect. Default is ``10``. The paramater to :class:`~airflow.providers.ssh.hooks.ssh.SSHHook` conn_timeout takes precedence.
```
Maybe we should document on the other side instead of here, like `remote_host`? Thoughts?

fatmumuhomer replied:
Sorry - what do you mean by "on the other side"?

jedcunningham replied:
My bad, I meant on the hook/operator side like was suggested for the operator in my other comment.

jedcunningham reviewed (changes requested):
Spotted a couple more things @fatmumuhomer. I think we are getting close now 👍

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Fix issue with defaults for timeouts in SSHHook/SSHOperator; add parametrized unit tests for SSHHook

fatmumuhomer committed:
Fix issue with defaults for timeouts in SSHHook/SSHOperator; add parametrized unit tests for SSHHook

fatmumuhomer committed:
Fix spelling mistake in ssh.rst

fatmumuhomer committed:
Fix spelling mistake in ssh.rst

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Move details for conn_timeout to SSHHook docstring instead of ssh.rst

fatmumuhomer committed:
Move details for conn_timeout to SSHHook docstring instead of ssh.rst

jedcunningham committed:
Chart: Make cleanup cronjob cmd/args configurable (#17970)

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

mik-laj committed:
Delete unnecessary parameters in EKSPodOperator (#17960)

josh-fell committed:
Remove default_args pattern + added get_current_context() use for Core Airflow example DAGs (#16866)

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>

potiuk committed:
Fix constraint generation properly (#17964)

The #17939 did not fix the problem finally. It turned out that
one more change was needed - since we now always upgrade to latest
dependencies in `push` and `schedule` type of build we do not need
to check for the variable UPGRADE_TO_NEWER_DEPENDENCIES (which
was not set in "Build Image" step.

This fixes it, but also changes the constraint generation to add
comments in the generated constraint files, describing how and
why the files are generated.

ephraimbuddy committed:
Fix max_active_runs not allowing moving of queued dagruns to running (#17945)

Currently, if you set max_active_runs for a dag and that dag has many queued dagruns
with execution dates older than another dag's queued dagruns, airflow will not move
the newer queued dagruns to running with the effect that only one dagruns would be in running at any time

This PR fixes this by updating the DagRun.last_scheduling_decision whenever a decision of scheduling
was made

bbenshalom committed:
Add DAG run endpoint for marking a dagrun success or failed(#17839)


Co-authored-by: bbenshalom <bbenshalom@outbrain.com>
Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>
Co-authored-by: Ephraim Anierobi <splendidzigy24@gmail.com>

bbovenzi committed:
Fix blank dag dependencies view (#17990)

* Fix blank dag dependencies view

* calculate graph if node and edges are empty

BasPH committed:
Serialize the template_ext attribute to show it in UI (#17985)

Co-authored-by: Bas Harenslak <bas@astronomer.io>

fatmumuhomer committed:
Merge branch 'apache:main' into airflow-16364

fatmumuhomer committed:
Remove commented out line

fatmumuhomer committed:
Remove commented out line

fatmumuhomer committed:
Merge branch 'airflow-16364' of https://github.com/fatmumuhomer/airflow into airflow-16364-v2

jedcunningham commented on tests/providers/ssh/operators/test_ssh.py:
```diff
@@ -138,7 +160,9 @@ def test_no_output_command(self):
     def test_arg_checking(self):
```
I'm wondering if we should split this apart? We now have 7 different cases in a single test.

fatmumuhomer replied:
If I understand what github is telling me, it looks like I can't be the one to resolve the conflict.

If I have some time tonight I'll split this test apart. I'm also happy to open a new PR for it if we want to close this one - whatever is easier for you all.

jedcunningham replied:
For the conflict, you should be able to resolve it by rebasing on the current main branch.

fatmumuhomer replied:
Alright, I **think** it's good now. I rebased and merged my changes to the file back in and there are no more conflicts. I hope I followed the right process - if not please let me know!

jedcunningham reviewed (approved):
There is one test that I think might be worth splitting up and there is a conflict now, otherwise LGTM.

fatmumuhomer committed:
Merge branch 'airflow-16364' of https://github.com/fatmumuhomer/airflow into airflow-16364-v2

fatmumuhomer committed:
Merge branch 'main' of https://github.com/apache/airflow into airflow-16364-v2
